core/environment/environment.go

```diff
@@ -614,11 +761,46 @@ func (env *Environment) runTasksAsHooks(hooksToTrigger task.Tasks) (errorMap map
 }
 
 func (env *Environment) TryTransition(t Transition) (err error) {
+	the.EventWriterWithTopic(topic.Environment).WriteEvent(&pb.Ev_EnvironmentEvent{
+		EnvironmentId:    env.id.String(),
+		State:            env.Sm.Current(),
+		CurrentRunNumber: env.currentRunNumber,
+		Message:          "transition starting",
+		Transition:       t.eventName(),
+	})
+
 	err = t.check()
 	if err != nil {
+		the.EventWriterWithTopic(topic.Environment).WriteEvent(&pb.Ev_EnvironmentEvent{
+			EnvironmentId:    env.id.String(),
+			State:            env.Sm.Current(),
+			CurrentRunNumber: env.currentRunNumber,
+			Error:            err.Error(),
+			Message:          "transition impossible",
+			Transition:       t.eventName(),
+		})
 		return
 	}
 	err = env.Sm.Event(context.Background(), t.eventName(), t)
+
+	if err != nil {
+		the.EventWriterWithTopic(topic.Environment).WriteEvent(&pb.Ev_EnvironmentEvent{
+			EnvironmentId:    env.id.String(),
+			State:            env.Sm.Current(),
+			CurrentRunNumber: env.currentRunNumber,
+			Error:            err.Error(),
+			Message:          "transition error",
+			Transition:       t.eventName(),
+		})
+	} else {
+		the.EventWriterWithTopic(topic.Environment).WriteEvent(&pb.Ev_EnvironmentEvent{
+			EnvironmentId:    env.id.String(),
+			State:            env.Sm.Current(),
+			CurrentRunNumber: env.currentRunNumber,
+			Message:          "transition completed successfully",
+			Transition:       t.eventName(),
+		})
+	}
 	return
```

Comment:
I'm not sure what you mean. We do need to define the environment events we're about to send, and while some fields are the same others change from one event to the other. Whether the fields are arguments to a function or fields of a struct doesn't change the fact that we need to provide them. Maybe I'm missing what you're proposing, can you elaborate?

Reply:
I am not proposing changes of functionality, I just see that you are creating a lot of `pb.Ev_EnvironmentEvent` messages in place with a lot of same parameters, the only changes are for parameters `Message` and `Error` ... So it would be great to create a function (eg. `createEV_Environment`) that would do it. and you would call it just like this: `createEv_Environment(env, error, message)`. This would reduce the bloat of the function significantly.